src/metabase/lib/stage.cljc

```diff
@@ -154,10 +154,14 @@
    unique-name-fn  :- fn?]
   (not-empty
    (for [expression (lib.expression/expressions query stage-number)]
-     (assoc expression
-            :lib/source               :source/expressions
-            :lib/source-column-alias  (:name expression)
-            :lib/desired-column-alias (unique-name-fn (:name expression))))))
+     (let [base-type (:base-type expression)]
+       (cond-> (assoc expression
+                      :lib/source               :source/expressions
+                      :lib/source-column-alias  (:name expression)
+                      :lib/desired-column-alias (unique-name-fn (:name expression)))
+         (and (not (:effective-type expression))
+              base-type)
+         (assoc :effective-type base-type))))))
```

Comment:
My understanding is that we don't want the FE to rely on `base-type`, `effective-type` should always be defined.